### PR TITLE
Create RCF Studio

### PR DESCRIPTION
**Pull request by luis-lin**
Create new RCF studio with sample inputs.

### CODE REVIEW

**Comment by luis-lin** [comment]
Demo

https://github.com/arista-netdevops-community/CloudVisionPortal-Examples/assets/143632211/f2a7220f-fec0-47a7-9c59-1aed4af811dd

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 1]
```diff
@@ -0,0 +1,178 @@
+- service: arista.studio.v1.StudioConfigService
+  method: Set
+  body:
+    value:
+      key:
+        studio_id: 7a161847-e603-49e5-904d-dc02fbef0d4e
+        workspace_id: ''
+      display_name: RCF v4.2
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:8]: Routing Control Functions (RCF)

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+      description: Code Units Management Tool
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:9]: RCF code unit management

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+      template:
+        type: TEMPLATE_TYPE_MAKO
+        body: |+
+          <%
+          # If the developer wants to see the container details of the script. 
+          # Defult debug setting is False. The info can be seen in `view build details` after clicking `review Workspace`
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:15]: Default (typo)

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+          debug = False
+
+          def output(content: str, indent=3, level=0):
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:18]: def applyIndentation

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+              '''
+              return: format multiline strings
+              '''
+              contentList = content.splitlines()
+              res = ''
+              for l in contentList:
+                  indentation = ' ' * (indent * level)
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:25]: move this out of the loop

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                  newL = indentation + l + '\n'
+                  res += newL
+              return res.rstrip()
+
+          codeUnitsData = {}
+          for codeUnit in codeUnits:
+              codeUnitsData.update({codeUnit['unitName']:output(codeUnit['unitContent'], level=2)})
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:32]: use dict comprehension directly which is more efficient.
```
codeUnitsData = { codeUnit[ 'unitName' ]: output( ... ) for codeUnit in codeUnits }
```

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+          deviceToUnits = {}
+          for assign in deviceAssignment.inputs:
+              codeUnitsList = assign['inputs']['value']['assignedUnits']
+              device_id = assign['devices'][0]
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:37]: formatting: camelCase

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+              deviceToUnits.update({device_id:codeUnitsList})
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:38]: use dict comprehension

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+          %>
+
+          %if debug:
+          ! ${codeUnitsData}
+          ! ${deviceToUnits}
+          %endif
+
+          %if ctx.getDevice().id in deviceToUnits.keys() and len(deviceToUnits[ctx.getDevice().id]) > 0:
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:46]: don't need `.keys()`. Python does a search in the keys already when we say `x in y`

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+          %   endfor 
+          %endif
+
+      input_schema:
+        fields:
+          values:
+            71bac242-81d9-4277-9789-3aa12259a742:
+              id: 71bac242-81d9-4277-9789-3aa12259a742
+              name: unitName
+              label: Code Unit Name
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:62]: Can we remove this?

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+              description: Code Unit Name
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:63]: not useful. let's remove the description

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+                static_options: null
+                format: null
+                length: null
+                pattern: null
+                dynamic_options: null
+                extra_values_allowed: false
+            3553abca-6b79-47e8-99e0-ea39a862172f:
+              id: 3553abca-6b79-47e8-99e0-ea39a862172f
+              name: unitContent
+              label: Code Unit Content
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:78]: Code Unit

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+              description: Routing Control Functions
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:79]: Provide the RCF code unit's configuration here. The code unit can have 1 or more RCF functions.

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+                static_options: null
+                format: null
+                length: null
+                pattern: null
+                dynamic_options: null
+                extra_values_allowed: false
+            ad2e5bbf-d3dd-461a-8d25-4faf3c699e12:
+              id: ad2e5bbf-d3dd-461a-8d25-4faf3c699e12
+              name: value
+              label: Code Units Group
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:94]: Code Unit and Content

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+              description: Group of members for Code Units
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:95]: If this doesn't appear anywhere, remove it.

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+              group_props:
+                members:
+                  values:
+                    - 71bac242-81d9-4277-9789-3aa12259a742
+                    - 3553abca-6b79-47e8-99e0-ea39a862172f
+            a4c1199e-ced0-4ed4-80b6-b6900f03f905:
+              id: a4c1199e-ced0-4ed4-80b6-b6900f03f905
+              name: codeUnits
+              label: Code Units
+              description: Code Unit Collections
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:107]: Define all code units used across all devices

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+              required: false
+              type: INPUT_FIELD_TYPE_COLLECTION
+              collection_props:
+                base_field_id: ad2e5bbf-d3dd-461a-8d25-4faf3c699e12
+                key: 71bac242-81d9-4277-9789-3aa12259a742
+            fefcf5bc-5db6-4273-9d37-733267a0429d:
+              id: fefcf5bc-5db6-4273-9d37-733267a0429d
+              name: assignedUnitName
+              label: Assign Code Units
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:116]: Code Units

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+                length: null
+                pattern: null
+                dynamic_options:
+                  values:
+                    - '{"fieldId":"71bac242-81d9-4277-9789-3aa12259a742"}'
+                extra_values_allowed: false
+            eea321af-f57c-43b5-9396-c1c3fbeaabb9:
+              id: eea321af-f57c-43b5-9396-c1c3fbeaabb9
+              name: assignedUnits
+              label: Assign Code Units Collection
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:134]: Assigned Code Units

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+                format: null
+                length: null
+                pattern: null
+                dynamic_options:
+                  values:
+                    - '{"fieldId":"71bac242-81d9-4277-9789-3aa12259a742"}'
+                extra_values_allowed: false
+            eea321af-f57c-43b5-9396-c1c3fbeaabb9:
+              id: eea321af-f57c-43b5-9396-c1c3fbeaabb9
+              name: assignedUnits
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:133]: devAssignedUnits

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+              description: ''
+              required: false
+              type: INPUT_FIELD_TYPE_COLLECTION
+              collection_props:
+                base_field_id: fefcf5bc-5db6-4273-9d37-733267a0429d
+                key: null
+            f035492e-131f-4378-bb57-678ad28a1b5d:
+              id: f035492e-131f-4378-bb57-678ad28a1b5d
+              name: value
+              label: Device Assignment Group
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:144]: Device Assignments

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+              label: Assign Code Units Collection
+              description: ''
+              required: false
+              type: INPUT_FIELD_TYPE_COLLECTION
+              collection_props:
+                base_field_id: fefcf5bc-5db6-4273-9d37-733267a0429d
+                key: null
+            f035492e-131f-4378-bb57-678ad28a1b5d:
+              id: f035492e-131f-4378-bb57-678ad28a1b5d
+              name: value
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:143]: devAssignments

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+              label: Device Assignment Group
+              description: Group of members for Unit Assignment
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:145]: Assignments to a single devcie

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+              required: false
+              type: INPUT_FIELD_TYPE_GROUP
+              group_props:
+                members:
+                  values:
+                    - eea321af-f57c-43b5-9396-c1c3fbeaabb9
+            2e151404-c66a-424e-bff8-ca912e09ad7b:
+              id: 2e151404-c66a-424e-bff8-ca912e09ad7b
+              name: deviceAssignment
+              label: Device Assignment
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:155]: Device Code Units

File: CloudVision_Studios/RCF Studio/RCF v4.2.yaml [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+              description: Choose code units from the drop-down menu to assign them to one or more devices.
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/RCF v4.2.yaml:156]: Assign multiple code units to devices

File: CloudVision_Studios/RCF Studio/Readme.md [hunk 25]
```diff
@@ -0,0 +1,2 @@
+RCF (Routing Control Function) Studio is a tool designed for the management of code units across multiple devices.
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/Readme.md:1]: Routing Control Function (RCF) Studio allows defining multiple code units and use them across devices in the network.

The general RCF language structure is documented here: https://www.arista.com/en/support/toi/eos-4-27-2f/15102-routing-control-functions-language-and-configuration
All attributes and builtin functions are documented here: https://www.arista.com/en/support/toi/eos-4-27-2f/15101-routing-control-functions-attribute-and-builtin-function-reference

File: CloudVision_Studios/RCF Studio/Samples_Inputs_RCF v4.2.yaml [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+      unitName: COMMON
+    - unitContent: |
+        function AS_65535_IN() {
+           # Rejects default route, accepts all other routes
+           if DEFAULT_ROUTE() {
+           return false;
+           }
+           return true;
+        }
+
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/Samples_Inputs_RCF v4.2.yaml:17]: one empty line

File: CloudVision_Studios/RCF Studio/Samples_Inputs_RCF v4.2.yaml [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+    - unitContent: |
+        function ACCEPT_ALL() {
+          # Accepts all routes
+          return true;
+        }
+
+        function DENY_ALL() {
+          # Rejects all routes
+          return false;
+        }
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/Samples_Inputs_RCF v4.2.yaml:34]: move this to "COMMON" code unit

File: CloudVision_Studios/RCF Studio/Samples_Inputs_RCF v4.2.yaml [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+      unitName: MAP_RULE
+    - unitContent: |
+        function CHECK_MED() {
+           if med is 10 {
+           return true;
```
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/Samples_Inputs_RCF v4.2.yaml:39]: fix indentation

File: CloudVision_Studios/RCF Studio/Samples_Inputs_RCF v4.2.yaml (hunk 27 above)
Comment (sbirmiwal-arista) [inline CloudVision_Studios/RCF Studio/Samples_Inputs_RCF v4.2.yaml:34]: let's use 3 whitespaces for indentation

**Review by sbirmiwal-arista — COMMENTED** [review]
Thank you for making the changes. Some comments attached. They are mostly cosmetic to make sure we use the right language. I'll let others chime in as well.

**Pull request merged**